<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>ABDB.Database_interface.PDB_details</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="ABDB-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

      <th class="navbar" width="100%"></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="ABDB-module.html">Package&nbsp;ABDB</a> ::
        <a href="ABDB.Database_interface-module.html">Module&nbsp;Database_interface</a> ::
        Class&nbsp;PDB_details
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript:void(0);" class="privatelink"
    onclick="toggle_private();">hide&nbsp;private</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<!-- ==================== CLASS DESCRIPTION ==================== -->
<h1 class="epydoc">Class PDB_details</h1><p class="nomargin-top"><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details">source&nbsp;code</a></span></p>
<pre class="base-tree">
object --+
         |
        <strong class="uidshort">PDB_details</strong>
</pre>

<hr />
<p>A class to hold the details about a pdb structure. This includes 
  information about the antibody and antigen molecules in the 
  structure.</p>

<!-- ==================== INSTANCE METHODS ==================== -->
<a name="section-InstanceMethods"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Instance Methods</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-InstanceMethods"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#__init__" class="summary-sig-name">__init__</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">id</span>,
        <span class="summary-sig-arg">database</span>=<span class="summary-sig-default">None</span>)</span><br />
      x.__init__(...) initializes x; see help(type(x)) for signature</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.__init__">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#__repr__" class="summary-sig-name">__repr__</a>(<span class="summary-sig-arg">self</span>)</span><br />
      repr(x)</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.__repr__">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#__getitem__" class="summary-sig-name">__getitem__</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">x</span>)</span><br />
      Get item can be used to return an Fab object from the object</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.__getitem__">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="__iter__"></a><span class="summary-sig-name">__iter__</span>(<span class="summary-sig-arg">self</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.__iter__">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_method"></a><span class="summary-sig-name">_set_method</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">method</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_method">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_resolution"></a><span class="summary-sig-name">_set_resolution</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">resolution</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_resolution">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_compound"></a><span class="summary-sig-name">_set_compound</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">compound</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_compound">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_r_free"></a><span class="summary-sig-name">_set_r_free</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">r_free</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_r_free">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_r_factor"></a><span class="summary-sig-name">_set_r_factor</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">r_factor</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_r_factor">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#_set_organism" class="summary-sig-name" onclick="show_private();">_set_organism</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">organism</span>=<span class="summary-sig-default">None</span>)</span><br />
      Set the organism of the pdb - this is taken from the pdb.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_organism">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#_set_species" class="summary-sig-name" onclick="show_private();">_set_species</a>(<span class="summary-sig-arg">self</span>)</span><br />
      The species field is the species of the FABS in the pdb.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_species">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#_set_common_species" class="summary-sig-name" onclick="show_private();">_set_common_species</a>(<span class="summary-sig-arg">self</span>)</span><br />
      The species field is the species of the FABS in the pdb.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_common_species">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_antigen_species"></a><span class="summary-sig-name">_set_antigen_species</span>(<span class="summary-sig-arg">self</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_antigen_species">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_short_header"></a><span class="summary-sig-name">_set_short_header</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">short_header</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_short_header">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_authors"></a><span class="summary-sig-name">_set_authors</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">authors</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_authors">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_date"></a><span class="summary-sig-name">_set_date</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">date</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_date">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_filepath"></a><span class="summary-sig-name">_set_filepath</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">filepath</span>=<span class="summary-sig-default">None</span>)</span><br />
      Set the filepath to the PDB structure.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_filepath">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_headerpath"></a><span class="summary-sig-name">_set_headerpath</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">headerpath</span>=<span class="summary-sig-default">None</span>)</span><br />
      Set the path to the header file for the structure</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_headerpath">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#_set_annotationpath" class="summary-sig-name" onclick="show_private();">_set_annotationpath</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">annotationpath</span>=<span class="summary-sig-default">None</span>)</span><br />
      Set the path to the directory containing antibody numbering 
      annotations.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_annotationpath">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#_set_sequencepath" class="summary-sig-name" onclick="show_private();">_set_sequencepath</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">sequencepath</span>=<span class="summary-sig-default">None</span>)</span><br />
      Set the path to the directory containing sequences for the various 
      chains in the structure.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_sequencepath">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_imgtpath"></a><span class="summary-sig-name">_set_imgtpath</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">imgtpath</span>=<span class="summary-sig-default">None</span>)</span><br />
      Set the path to the directory containing imgt gene annotations for 
      the chains.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_imgtpath">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_abanglepath"></a><span class="summary-sig-name">_set_abanglepath</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">abanglepath</span>=<span class="summary-sig-default">None</span>)</span><br />
      Set the path to the directory containing imgt gene annotations for 
      the chains.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_abanglepath">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_summarypath"></a><span class="summary-sig-name">_set_summarypath</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">summarypath</span>=<span class="summary-sig-default">None</span>)</span><br />
      Set the path to the directory containing summary file for the entry</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_summarypath">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_clusteringpath"></a><span class="summary-sig-name">_set_clusteringpath</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">clusteringpath</span>=<span class="summary-sig-default">None</span>)</span><br />
      Set the path to the directory containing information about the cdr 
      clusters</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_clusteringpath">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_inhouse"></a><span class="summary-sig-name">_set_inhouse</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">inhouse</span>=<span class="summary-sig-default">False</span>)</span><br />
      Set whether the structure is an inhouse structure or not.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_inhouse">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#_set_fabs" class="summary-sig-name" onclick="show_private();">_set_fabs</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">fabs</span>=<span class="summary-sig-default">None</span>)</span><br />
      Set the fabs in the structure.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_fabs">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_header"></a><span class="summary-sig-name">_set_header</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Read the header from the database and set it as an attribute.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_header">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#_set_numbering" class="summary-sig-name" onclick="show_private();">_set_numbering</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">scheme</span>=<span class="summary-sig-default">None</span>)</span><br />
      Set the antibody numbering for the chains as stored in the database.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_numbering">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#_set_alignment" class="summary-sig-name" onclick="show_private();">_set_alignment</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">scheme</span>=<span class="summary-sig-default">None</span>)</span><br />
      An alignment is needed between the seqres sequence and the structural
      sequence (missing residues in the structure will lead to 
      differences).</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_alignment">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_sequence"></a><span class="summary-sig-name">_set_sequence</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">scheme</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_sequence">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_raw_sequences"></a><span class="summary-sig-name">_set_raw_sequences</span>(<span class="summary-sig-arg">self</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_raw_sequences">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_antigens"></a><span class="summary-sig-name">_set_antigens</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">antigens</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_antigens">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_complex"></a><span class="summary-sig-name">_set_complex</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">complex</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_complex">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_scfv"></a><span class="summary-sig-name">_set_scfv</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">scfv</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_scfv">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_completefab"></a><span class="summary-sig-name">_set_completefab</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">completefab</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_completefab">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_contains_completefab"></a><span class="summary-sig-name">_set_contains_completefab</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">contains_completefab</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_contains_completefab">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_engineered"></a><span class="summary-sig-name">_set_engineered</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">engineered</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_engineered">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#_set_affinity" class="summary-sig-name" onclick="show_private();">_set_affinity</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">affinity</span>=<span class="summary-sig-default">None</span>)</span><br />
      Affinity data from Jin's database.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_affinity">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#_set_delta_g" class="summary-sig-name" onclick="show_private();">_set_delta_g</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">delta_g</span>=<span class="summary-sig-default">None</span>)</span><br />
      Delta G data from Jin's database.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_delta_g">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#_set_pmid" class="summary-sig-name" onclick="show_private();">_set_pmid</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">pmid</span>=<span class="summary-sig-default">None</span>)</span><br />
      Pubmed id of the relevent publication - currently only for structures
      with affinity data</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_pmid">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_affinity_method"></a><span class="summary-sig-name">_set_affinity_method</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">affinity_method</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_affinity_method">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_temperature"></a><span class="summary-sig-name">_set_temperature</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">temperature</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_temperature">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_imgt_details"></a><span class="summary-sig-name">_set_imgt_details</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Set the imgt details of all the antibody chains in the pdb.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_imgt_details">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_orientation_angles"></a><span class="summary-sig-name">_set_orientation_angles</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Read the orientation angles from the database.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_orientation_angles">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_set_summary_lines"></a><span class="summary-sig-name">_set_summary_lines</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Read the summary lines from the database.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_summary_lines">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#_load_structure" class="summary-sig-name" onclick="show_private();">_load_structure</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">scheme</span>=<span class="summary-sig-default">None</span>)</span><br />
      Load the structure from the Database.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._load_structure">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#clear_structure" class="summary-sig-name">clear_structure</a>(<span class="summary-sig-arg">self</span>)</span><br />
      This deletes the structure attribute if it has been loaded.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.clear_structure">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_method"></a><span class="summary-sig-name">get_method</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the experimental method used to obtain the structure.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_method">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_resolution" class="summary-sig-name">get_resolution</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the resolution of the structure.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_resolution">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_r_factor" class="summary-sig-name">get_r_factor</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the r_factor (observed) of the structure.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_r_factor">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_r_free" class="summary-sig-name">get_r_free</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the r_free of the structure.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_r_free">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_header"></a><span class="summary-sig-name">get_header</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the header of the PDB file</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_header">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_short_header" class="summary-sig-name">get_short_header</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the short header of the PDB file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_short_header">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_authors"></a><span class="summary-sig-name">get_authors</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the authors of the structure.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_authors">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_date"></a><span class="summary-sig-name">get_date</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the date deposited in the PDB.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_date">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_pmid" class="summary-sig-name">get_pmid</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the pubmed id of the structure.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_pmid">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_compound"></a><span class="summary-sig-name">get_compound</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the name of the compound</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_compound">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_organism" class="summary-sig-name">get_organism</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the organism(s) the molecule has been obtained from.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_organism">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_species" class="summary-sig-name">get_species</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the species of the antibodies chains in the molecule.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_species">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_common_species" class="summary-sig-name">get_common_species</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the species of the antibodies chains in the molecule.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_common_species">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_antigen_species" class="summary-sig-name">get_antigen_species</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the species of the antigen molecule if present.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_antigen_species">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_filepath"></a><span class="summary-sig-name">get_filepath</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the path to the PDB structure in the database.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_filepath">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_fabs" class="summary-sig-name">get_fabs</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get a list of the fabs (or unpaired antibody chains) in the PDB 
      structure.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_fabs">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_numbering" class="summary-sig-name">get_numbering</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">scheme</span>=<span class="summary-sig-default">None</span>)</span><br />
      Get the numbering of the chains in the PDB file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_numbering">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_alignment" class="summary-sig-name">get_alignment</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">scheme</span>=<span class="summary-sig-default">None</span>)</span><br />
      Get the alignment dictionary between the seqres sequence and the 
      structure sequence.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_alignment">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_structure" class="summary-sig-name">get_structure</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">scheme</span>=<span class="summary-sig-default">None</span>)</span><br />
      Get the structure loaded as a biopython pdb structure put into an 
      &quot;antibody context&quot;.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_structure">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_sequence" class="summary-sig-name">get_sequence</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">scheme</span>=<span class="summary-sig-default">None</span>)</span><br />
      Get the sequences of the antibody chains in the structure.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_sequence">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_raw_sequences"></a><span class="summary-sig-name">get_raw_sequences</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Get a dictionary containing the chain identifier as the key and the 
      sequence as the value.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_raw_sequences">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_summary_lines" class="summary-sig-name">get_summary_lines</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the summary lines for the structure.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_summary_lines">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_antigens" class="summary-sig-name">get_antigens</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get a list of the antigen molecules in the PDB file.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_antigens">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_light_chain_types"></a><span class="summary-sig-name">get_light_chain_types</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Get a list of the light chain types of the fabs in the pdb (normally 
      these will all be the same)</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_light_chain_types">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_light_subclasses"></a><span class="summary-sig-name">get_light_subclasses</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Get a list of the light subclass of the fabs in the pdb (normally 
      these will all be the same)</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_light_subclasses">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_heavy_subclasses"></a><span class="summary-sig-name">get_heavy_subclasses</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Get a list of the light subclass of the fabs in the pdb (normally 
      these will all be the same)</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_heavy_subclasses">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_imgt_details" class="summary-sig-name">get_imgt_details</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get a dictionary containing the imgt details for all the antibody 
      chains in this pdb.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_imgt_details">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_orientation_angles"></a><span class="summary-sig-name">get_orientation_angles</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Get a dictionary containing the abangle angles for each fv in this 
      pdb.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_orientation_angles">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#is_complex" class="summary-sig-name">is_complex</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Check whether the Fabs in the structure are bound to antigens or not.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.is_complex">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type"><code>float</code> or None</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_affinity" class="summary-sig-name">get_affinity</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the affinity (Kd) of the antibody in this structure to its 
      antigen.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_affinity">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type"><code>float</code> or None</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#get_delta_g" class="summary-sig-name">get_delta_g</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the delta G value for the antibody in this structure to its 
      antigen.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_delta_g">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_temperature"></a><span class="summary-sig-name">get_temperature</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the temperature at which the affinity information was obtained.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_temperature">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_affinity_method"></a><span class="summary-sig-name">get_affinity_method</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Get the method by which the affinity information was obtained.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_affinity_method">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#is_scfv" class="summary-sig-name">is_scfv</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Check whether any of the Fabs in the structure are single chain fvs.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.is_scfv">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#is_completefab" class="summary-sig-name">is_completefab</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Check whether the fvs in the structure are complete i.e.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.is_completefab">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#has_completefab" class="summary-sig-name">has_completefab</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Check whether the pdb contains a paired VH-VL.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.has_completefab">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ABDB.Database_interface.PDB_details-class.html#is_engineered" class="summary-sig-name">is_engineered</a>(<span class="summary-sig-arg">self</span>)</span><br />
      Check whether the structure is engineered or not.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.is_engineered">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="is_imgt_from_imgt"></a><span class="summary-sig-name">is_imgt_from_imgt</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Check whether source of the imgt details is from the imgt webserver.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.is_imgt_from_imgt">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="is_imgt_from_ABDB"></a><span class="summary-sig-name">is_imgt_from_ABDB</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Check whether source of the imgt details is assigned by ABDB.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.is_imgt_from_ABDB">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="is_inhouse"></a><span class="summary-sig-name">is_inhouse</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Is the entry an in-house structure?</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.is_inhouse">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
  <tr>
    <td colspan="2" class="summary">
    <p class="indent-wrapped-lines"><b>Inherited from <code>object</code></b>:
      <code>__delattr__</code>,
      <code>__format__</code>,
      <code>__getattribute__</code>,
      <code>__hash__</code>,
      <code>__new__</code>,
      <code>__reduce__</code>,
      <code>__reduce_ex__</code>,
      <code>__setattr__</code>,
      <code>__sizeof__</code>,
      <code>__str__</code>,
      <code>__subclasshook__</code>
      </p>
    </td>
  </tr>
</table>
<!-- ==================== PROPERTIES ==================== -->
<a name="section-Properties"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Properties</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Properties"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
  <tr>
    <td colspan="2" class="summary">
    <p class="indent-wrapped-lines"><b>Inherited from <code>object</code></b>:
      <code>__class__</code>
      </p>
    </td>
  </tr>
</table>
<!-- ==================== METHOD DETAILS ==================== -->
<a name="section-MethodDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Method Details</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-MethodDetails"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
</table>
<a name="__init__"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">__init__</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">id</span>,
        <span class="sig-arg">database</span>=<span class="sig-default">None</span>)</span>
    <br /><em class="fname">(Constructor)</em>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.__init__">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>x.__init__(...) initializes x; see help(type(x)) for signature</p>
  <dl class="fields">
    <dt>Overrides:
        object.__init__
        <dd><em class="note">(inherited documentation)</em></dd>
    </dt>
  </dl>
</td></tr></table>
</div>
<a name="__repr__"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">__repr__</span>(<span class="sig-arg">self</span>)</span>
    <br /><em class="fname">(Representation operator)</em>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.__repr__">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>repr(x)</p>
  <dl class="fields">
    <dt>Overrides:
        object.__repr__
        <dd><em class="note">(inherited documentation)</em></dd>
    </dt>
  </dl>
</td></tr></table>
</div>
<a name="__getitem__"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">__getitem__</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">x</span>)</span>
    <br /><em class="fname">(Indexing operator)</em>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.__getitem__">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get item can be used to return an Fab object from the object</p>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>x</code></strong> (str) - Heavy_ident+Light_ident</li>
    </ul></dd>
  </dl>
</td></tr></table>
</div>
<a name="_set_organism"></a>
<div class="private">
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">_set_organism</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">organism</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_organism">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Set the organism of the pdb - this is taken from the pdb. It is 
  formatted in like: unique_organism1;  unique_organism2; unique_organism3 
  if there are three different unique organisms in the pdb.</p>
  <p>If a molecule is made of more than one organism (e.g. a chimera) the 
  format would be: organism1a, organism1b;  unique_organism2; 
  unique_organism3</p>
  <p>So we split the organism field into an internal variable and assign it
  to the antigen and fab molecules. (they can see _organisms)</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="_set_species"></a>
<div class="private">
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">_set_species</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_species">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>The species field is the species of the FABS in the pdb. This is NOT 
  the same as the organism field which is the raw pdb field. In the rare 
  case that there are fabs of different species in the same pdb, these two 
  species will be separated by &quot;; &quot;. Chimeric fabs will be 
  labelled &quot;CHIMERIC species1/species2 /... &quot;</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="_set_common_species"></a>
<div class="private">
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">_set_common_species</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_common_species">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>The species field is the species of the FABS in the pdb. This is NOT 
  the same as the organism field which is the raw pdb field. In the rare 
  case that there are fabs of different species in the same pdb, these two 
  species will be separated by &quot;; &quot;. Chimeric fabs will be 
  labelled &quot;CHIMERIC species1/species2 /... &quot;</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="_set_annotationpath"></a>
<div class="private">
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">_set_annotationpath</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">annotationpath</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_annotationpath">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Set the path to the directory containing antibody numbering 
  annotations. These annotations have been generated by abnum. 
  http://www.bioinf.org.uk/abs/abnum/</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="_set_sequencepath"></a>
<div class="private">
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">_set_sequencepath</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">sequencepath</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_sequencepath">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Set the path to the directory containing sequences for the various 
  chains in the structure. These are fasta files.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="_set_fabs"></a>
<div class="private">
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">_set_fabs</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">fabs</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_fabs">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Set the fabs in the structure. fabs is a list of dictionaries 
  containing information about each &quot;fab&quot;. An fab might be a 
  single chain if it has not been paired.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="_set_numbering"></a>
<div class="private">
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">_set_numbering</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">scheme</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_numbering">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Set the antibody numbering for the chains as stored in the database. 
  The numbering is done using the seqres sequence for each chain.</p>
  <p>Each annotation file describes the numbering of a variable region of a
  chain. Each chain may have more than one variable region (e.g. scfv)</p>
  <p>Annotations are stored for the pdb as:</p>
  <p>pdb_chain_region.ann</p>
  <p>e.g for 12e8 we have: 12e8_H_VH.ann 12e8_L_VL.ann 12e8_M_VL.ann 
  12e8_P_VH.ann</p>
  <p>These are Chothia numbered.</p>
  <p>self.numbering is a dictionary with chain identifiers as keys.</p>
  <p>For each chain there is a list of variable regions.</p>
  <p>Each entry of the list is a tuple of chain_numbering and 
  chain_type.</p>
  <p>chain_type will be 'H' or 'L'.</p>
  <p>chain numbering is a list of biopython.PDB residue identifiers and one
  letter amino acid codes.</p>
  <p>e.g. [ ( (1,' '), E ), ( (2, ' '), Q ) ..... ]</p>
  <p>This is not pretty but it works... We may want to re-think this in the
  future.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="_set_alignment"></a>
<div class="private">
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">_set_alignment</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">scheme</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_alignment">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>An alignment is needed between the seqres sequence and the structural 
  sequence (missing residues in the structure will lead to 
  differences).</p>
  <p>This retrieves the alignment from the database</p>
  <p>We create an alignment dictionary between the seqres sequence and the 
  structure sequence.</p>
  <p>See ABDB.Annotate.annotate.get_alignment_dict for more details.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="_set_affinity"></a>
<div class="private">
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">_set_affinity</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">affinity</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_affinity">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Affinity data from Jin's database.</p>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>affinity</code></strong> (<code>float</code> or <code>str</code> or None. Expect an 
          exponent format for the string or float. e.g 1.0e-10) - The affinity data collected (Kd) - or None</li>
    </ul></dd>
  </dl>
</td></tr></table>
</div>
<a name="_set_delta_g"></a>
<div class="private">
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">_set_delta_g</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">delta_g</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_delta_g">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Delta G data from Jin's database.</p>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>delt_g</code></strong> - The calculated delta G value</li>
    </ul></dd>
  </dl>
</td></tr></table>
</div>
<a name="_set_pmid"></a>
<div class="private">
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">_set_pmid</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">pmid</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._set_pmid">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Pubmed id of the relevent publication - currently only for structures 
  with affinity data</p>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>pmid</code></strong> - The pubmed id of the paper</li>
    </ul></dd>
  </dl>
</td></tr></table>
</div>
<a name="_load_structure"></a>
<div class="private">
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">_load_structure</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">scheme</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details._load_structure">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Load the structure from the Database. This sets the structure 
  attribute. Use get_structure to return the structure.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="clear_structure"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">clear_structure</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.clear_structure">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <pre class="literalblock">

This deletes the structure attribute if it has been loaded.
Note: if you have other references to the structure object they MUST ALSO BE DELETED to free up the memory.
e.g. s = pdb_details.get_structure()
     pdb_details.clear_structure()
     del s
     
For batch analysis of structures with this code this should be considered.
No memory considerations are made (i.e. do not load in ALL the structures at the same time or your machine will die :( )

</pre>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_resolution"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_resolution</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_resolution">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the resolution of the structure. This only applies to X-ray 
  structures.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_r_factor"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_r_factor</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_r_factor">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the r_factor (observed) of the structure. This only applies to 
  X-ray structures.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_r_free"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_r_free</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_r_free">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the r_free of the structure. This only applies to X-ray 
  structures.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_short_header"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_short_header</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_short_header">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the short header of the PDB file. This often contains the molecule
  name or title.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_pmid"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_pmid</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_pmid">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the pubmed id of the structure.</p>
  <p>Currently only returns pmid for structures with affinity data.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_organism"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_organism</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_organism">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the organism(s) the molecule has been obtained from. Note that 
  this is the raw organism field from the PDB. NOT the same as get_species.
  Therefore, those structures that are more than one species will have both
  species names in the field.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_species"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_species</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_species">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the species of the antibodies chains in the molecule. If the fabs 
  have different species (rare) then the seperate species will be returned 
  separated with a &quot;;&quot;. Chimeric antibodies should have two 
  species separated by a &quot;,&quot;.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_common_species"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_common_species</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_common_species">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the species of the antibodies chains in the molecule. If the fabs 
  have different species (rare) then the seperate species will be returned 
  separated with a &quot;;&quot;. Chimeric antibodies should have two 
  species separated by a &quot;,&quot;.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_antigen_species"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_antigen_species</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_antigen_species">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the species of the antigen molecule if present. This is only 
  defined for biological protein and peptide antigens.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_fabs"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_fabs</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_fabs">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get a list of the fabs (or unpaired antibody chains) in the PDB 
  structure. Each item of the list will be an fab_details object.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_numbering"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_numbering</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">scheme</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_numbering">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the numbering of the chains in the PDB file. See <a 
  href="ABDB.Database_interface.PDB_details-class.html#_set_numbering" 
  class="link" onclick="show_private();">set_numbering</a> for more 
  information about the numbering attribute.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_alignment"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_alignment</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">scheme</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_alignment">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the alignment dictionary between the seqres sequence and the 
  structure sequence.</p>
  <p>This is a dictionary with chain IDs for keys and alignment 
  dictionaries for values.</p>
  <p>Alignment dictionary tells you that the the i'th (key) position in the
  seqres sequence corresponds to the j'th (value) position in the structure
  sequence.</p>
  <p>See ABDB.Annotate.annotate.get_alignment_dict for more details.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_structure"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_structure</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">scheme</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_structure">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the structure loaded as a biopython pdb structure put into an 
  &quot;antibody context&quot;. The antibody chains will have been 
  renumbered using the Chothia numbering system. See the ABDB.AbPDB package
  for more information.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_sequence"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_sequence</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">scheme</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_sequence">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <pre class="literalblock">

Get the sequences of the antibody chains in the structure.

Returns a nested dictionary.

For example a PDB has antibody chains P and M. P is a heavy chain, M is a light chain.

s = pdb_object.get_sequence()

s.keys() # [ &quot;P&quot; , &quot;M&quot; ]

s[&quot;P&quot;] # This returns a dictionary for each variable region in the chain ( single chain fvs can have more than one variable region e.g. 2kh2).         

s[&quot;P&quot;][&quot;H&quot;] # The heavy region of chain P          
s[&quot;M&quot;][&quot;L&quot;] # The light region of chain M

s[&quot;P&quot;][&quot;H&quot;] # contains:
            #    structurefull    - the sequence found in the structure (i.e. resolved residues)
            #    seqresfull       - the sequence found in the seqres record from the pdb
            #    seqresregion: H  - the variable region of the seqres record ( non-variable region is represented by &quot;-&quot; )
            
s[&quot;M&quot;][&quot;L&quot;] # contains:
            #    structurefull    - the sequence found in the structure (i.e. resolved residues)
            #    seqresfull       - the sequence found in the seqres record from the pdb
            #    seqresregion: L  - the variable region of the seqres record ( non-variable region is represented by &quot;-&quot; )
  

</pre>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_summary_lines"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_summary_lines</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_summary_lines">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the summary lines for the structure. This will return a dictionary
  with either 'header' or a fab identifier ( i.e (pdb, H, L, modelid) ) as 
  keys and the line as a value.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_antigens"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_antigens</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_antigens">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get a list of the antigen molecules in the PDB file. Each item of the 
  list will be an antigen_details object. If there are no antigens this 
  will be an empty list.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_imgt_details"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_imgt_details</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_imgt_details">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get a dictionary containing the imgt details for all the antibody 
  chains in this pdb. Query individual fabs for their details.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="is_complex"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">is_complex</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.is_complex">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Check whether the Fabs in the structure are bound to antigens or not. 
  This is true if any of the fabs are bound and false otherwise. Inspect 
  each fab individually to check their bound state</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_affinity"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_affinity</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_affinity">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the affinity (Kd) of the antibody in this structure to its 
  antigen.</p>
  <dl class="fields">
    <dt>Returns: <code>float</code> or None</dt>
        <dd>The affinity (Kd) of the antibody to its antigen. Units are M.</dd>
  </dl>
</td></tr></table>
</div>
<a name="get_delta_g"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_delta_g</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.get_delta_g">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Get the delta G value for the antibody in this structure to its 
  antigen.</p>
  <dl class="fields">
    <dt>Returns: <code>float</code> or None</dt>
        <dd>The delta G value of the antibody to its antigen. Units are 
          kcal/mol</dd>
  </dl>
</td></tr></table>
</div>
<a name="is_scfv"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">is_scfv</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.is_scfv">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Check whether any of the Fabs in the structure are single chain fvs. 
  This is parsed from the header record of the PDB using nlp. If their are 
  multiple variable regions on the chain these will also be a scfv.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="is_completefab"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">is_completefab</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.is_completefab">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Check whether the fvs in the structure are complete i.e. paired. If 
  there are any unpaired antibody chains within the structure this returns 
  False, otherwise True. * set to be depreciated as is misleading - in 
  favour of has_completefab method</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="has_completefab"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">has_completefab</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.has_completefab">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Check whether the pdb contains a paired VH-VL. Returns True if it does
  False if not. Inspect each individual fab details object to see which 
  ones are paired.</p>
  <p>Slightly misleading.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="is_engineered"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">is_engineered</span>(<span class="sig-arg">self</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ABDB.Database_interface-pysrc.html#PDB_details.is_engineered">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Check whether the structure is engineered or not. This has been parsed
  from the header.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="ABDB-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

      <th class="navbar" width="100%"></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Fri Jul  3 12:03:46 2015
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
