{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import pandas as pd\n",
    "import copy\n",
    "import numpy as np\n",
    "import json\n",
    "import pickle\n",
    "import random\n",
    "import multiprocessing\n",
    "from numba import njit, jit\n",
    "\n",
    "Entrez.email = \"fabian.spoendlin@exeter.ox.ac.uk\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('CoV-AbDab_181021.csv', 'r') as f:\n",
    "    CovAbDab = pd.read_csv(f)   \n",
    "\n",
    "# missing seqs in Cov abdab are ND, this can be protein seq -> replace with something thats not protein seq\n",
    "CovAbDab['VH or VHH'].replace(to_replace='ND', value='no sequence available', inplace=True)\n",
    "CovAbDab['VL'].replace(to_replace='ND', value='no sequence available', inplace=True)\n",
    "CovAbDab['VH or VHH'].fillna('no sequence available', inplace=True)\n",
    "CovAbDab['VL'].fillna('no sequence available', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#search nucleotides and check how many of the sequences are in Covab Dab\n",
    "handle = Entrez.esearch(db='nucleotide', term='anti-sars-cov-2[All Fields] AND immunoglobulin[All Fields]', retmax='50')\n",
    "record = Entrez.read(handle)\n",
    "\n",
    "nucleotides = []\n",
    "\n",
    "for ID in record['IdList']:\n",
    "    nucelotide = Entrez.efetch(db=\"nucleotide\", id=ID, rettype=\"gb\", retmode=\"text\")\n",
    "    nucelotide_formated = SeqIO.read(nucelotide,'genbank')\n",
    "    nucleotides.append(nucelotide_formated)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17 50\n"
     ]
    }
   ],
   "source": [
    "summary_n = []\n",
    "\n",
    "for entry in nucleotides:\n",
    "\n",
    "    lenght_sequence = len(entry.seq)\n",
    "    remove_bases = lenght_sequence % 3\n",
    "    if remove_bases == 0:\n",
    "        aa_seq = str(entry.seq.translate()) # problems if nucleotide sequence is not in correct frame\n",
    "    else:\n",
    "        aa_seq = str(entry.seq[:-remove_bases].translate())\n",
    "\n",
    "    containes = []\n",
    "    for VH in CovAbDab['VH or VHH']:\n",
    "        if VH in aa_seq:\n",
    "            containes.append('heavy chain')\n",
    "    for VL in CovAbDab['VL']:\n",
    "        if str(VL) in aa_seq: # nan in VL that is formated as float\n",
    "            containes.append('light chain')\n",
    "    \n",
    "    if 'heavy chain' in containes:\n",
    "        summary_n.append('heavy chain')\n",
    "        \n",
    "    elif 'light chain' in containes:\n",
    "        summary_n.append('light chain') \n",
    "\n",
    "    else:\n",
    "        summary_n.append('not in covab-dab')\n",
    "        \n",
    "print(summary_n.count('not in covab-dab'),len(summary_n))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/Bio/GenBank/__init__.py:1143: BiopythonParserWarning: Dropping bond qualifier in feature location\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "#search proteins and check how many of the sequences are in Covab Dab\n",
    "handle = Entrez.esearch(db='protein', term='anti-sars-cov-2[All Fields] AND immunoglobulin[All Fields]', retmax='50')\n",
    "record = Entrez.read(handle)\n",
    "\n",
    "proteins = []\n",
    "\n",
    "for ID in record['IdList']:\n",
    "    protein = Entrez.efetch(db=\"protein\", id=ID, rettype=\"gb\", retmode=\"text\")\n",
    "    protein_formated = SeqIO.read(protein,'genbank')\n",
    "    proteins.append(protein_formated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13 50\n"
     ]
    }
   ],
   "source": [
    "summary_p = []\n",
    "\n",
    "for entry in proteins:\n",
    "\n",
    "    aa_seq = str(entry.seq)\n",
    "\n",
    "    containes = []\n",
    "    for VH in CovAbDab['VH or VHH']:\n",
    "        if VH in aa_seq:\n",
    "            containes.append('heavy chain')\n",
    "    for VL in CovAbDab['VL']:\n",
    "        if str(VL) in aa_seq: # nan in VL that is formated as float\n",
    "            containes.append('light chain')\n",
    "    \n",
    "    if 'heavy chain' in containes:\n",
    "        summary_p.append('heavy chain')\n",
    "        \n",
    "    elif 'light chain' in containes:\n",
    "        summary_p.append('light chain') \n",
    "\n",
    "    else:\n",
    "        summary_p.append('not in covab-dab')\n",
    "\n",
    "print(summary_p.count('not in covab-dab'),len(summary_p))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50 50\n",
      "50 50\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/Bio/Seq.py:2979: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "#nucelotide and protein database contains sequence of whole chain not just VL and VH, thus this comparison does not work\n",
    "summary_2_n = []\n",
    "\n",
    "for entry in nucleotides:\n",
    "\n",
    "    aa_seq = str(entry.seq.translate())\n",
    "   \n",
    "    if aa_seq in CovAbDab['VH or VHH']:\n",
    "        summary_2_n.append('heavy chain')\n",
    "\n",
    "    elif aa_seq in str(CovAbDab['VL']):\n",
    "        summary_2_n.append('light chain') \n",
    "\n",
    "    else:\n",
    "        summary_2_n.append('not in covab-dab')\n",
    "\n",
    "print(summary_2_n.count('not in covab-dab'),len(summary_2_n))\n",
    "\n",
    "\n",
    "summary_2_p = []\n",
    "\n",
    "for entry in proteins:\n",
    "\n",
    "    aa_seq = str(entry.seq)\n",
    "   \n",
    "    if aa_seq in CovAbDab['VH or VHH']:\n",
    "        summary_2_p.append('heavy chain')\n",
    "\n",
    "    elif aa_seq in str(CovAbDab['VL']):\n",
    "        summary_2_p.append('light chain') \n",
    "\n",
    "    else:\n",
    "        summary_2_p.append('not in covab-dab')\n",
    "\n",
    "print(summary_2_p.count('not in covab-dab'),len(summary_2_p))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Protein search pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of entries: 2685\n"
     ]
    }
   ],
   "source": [
    "# perform keyword search\n",
    "\n",
    "# specify the terms used for the search\n",
    "search = 'anti-sars-cov-2[All Fields] AND immunoglobulin[All Fields]'\n",
    "\n",
    "# search protein data base with keywords and find out how many entries are found\n",
    "handle = Entrez.esearch(db='protein', term='anti-sars-cov-2[All Fields] AND immunoglobulin[All Fields]', retmax='2')\n",
    "record = Entrez.read(handle)\n",
    "number_of_entries = int(record['Count'])\n",
    "print('number of entries:', number_of_entries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/1416774354.py:16: NumbaWarning: \n",
      "Compilation is falling back to object mode WITH looplifting enabled because Function \"search_in_covabdab\" failed type inference due to: non-precise type array(pyobject, 1d, C)\n",
      "During: typing of argument at /var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/1416774354.py (19)\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/1416774354.py\", line 19:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  @jit\n",
      "/var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/1416774354.py:16: NumbaWarning: \n",
      "Compilation is falling back to object mode WITHOUT looplifting enabled because Function \"search_in_covabdab\" failed type inference due to: Cannot determine Numba type of <class 'numba.core.dispatcher.LiftedLoop'>\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/1416774354.py\", line 22:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  @jit\n",
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/numba/core/object_mode_passes.py:151: NumbaWarning: Function \"search_in_covabdab\" was compiled in object mode without forceobj=True, but has lifted loops.\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/1416774354.py\", line 19:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  warnings.warn(errors.NumbaWarning(warn_msg,\n",
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/numba/core/object_mode_passes.py:161: NumbaDeprecationWarning: \n",
      "Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.\n",
      "\n",
      "For more information visit https://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/1416774354.py\", line 19:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  warnings.warn(errors.NumbaDeprecationWarning(msg,\n",
      "/var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/1416774354.py:16: NumbaWarning: \n",
      "Compilation is falling back to object mode WITHOUT looplifting enabled because Function \"search_in_covabdab\" failed type inference due to: non-precise type array(pyobject, 1d, C)\n",
      "During: typing of argument at /var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/1416774354.py (22)\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/1416774354.py\", line 22:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  @jit\n",
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/numba/core/object_mode_passes.py:151: NumbaWarning: Function \"search_in_covabdab\" was compiled in object mode without forceobj=True.\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/1416774354.py\", line 22:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  warnings.warn(errors.NumbaWarning(warn_msg,\n",
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/numba/core/object_mode_passes.py:161: NumbaDeprecationWarning: \n",
      "Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.\n",
      "\n",
      "For more information visit https://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/1416774354.py\", line 22:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  warnings.warn(errors.NumbaDeprecationWarning(msg,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total sequences assessed: 2685\n",
      "number of genbank sequences not in covab dab: 409\n",
      "number of genbank sequences found in covab dab: 2276\n",
      "match rate: 0.8476722532588454\n",
      "number of genebank sequences that have multiple matches in covab dab: 636.0\n",
      "number of genebank entries with non unique match in covab dab: 13\n",
      "-------------\n",
      "total sequences in covab dab: 4198\n",
      "number of Covab Dab VH sequences found: 1110\n",
      "number of Covab Dab VL sequences found: 1153\n",
      "number of Covab Dab VH VL pairings found: 1059\n",
      "number of Covab Dab entires where either VH or VL was found: 1204\n",
      "-------------\n",
      "percentage of covab dab entries with pairing found: 25.226298237255833\n",
      "percentage of covab dab entries with either VL or VH found: 28.680323963792283\n"
     ]
    }
   ],
   "source": [
    "# download all entries from search\n",
    "\n",
    "handle = Entrez.esearch(db='protein', term=search, retmax=number_of_entries)\n",
    "record = Entrez.read(handle)\n",
    "\n",
    "# 25 searches per second\n",
    "protein_handle =  Entrez.efetch(db=\"protein\", id=record['IdList'], rettype=\"gb\", retmode=\"xml\")\n",
    "proteins = Entrez.read(protein_handle)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# compare found sequences to covab dab\n",
    "\n",
    "# use numba to speed up computation\n",
    "@jit\n",
    "def search_in_covabdab(aa_seqs, VH_arr, VL_arr, VH_found, VL_found, sequences_not_in_covabdab):\n",
    "\n",
    "    sequences_not_in_covabdab = 0\n",
    "\n",
    "    # loop throught aa seqs\n",
    "    for aa_seq in aa_seqs:\n",
    "        sequence_found = False\n",
    "\n",
    "        # loop throught covab dab entries\n",
    "        for i in range(len(VH_arr)):\n",
    "\n",
    "            # in case VH is in covab dab increase the VH count of this entry by 1\n",
    "            if VH_arr[i] in aa_seq:\n",
    "                VH_found[i] = VH_found[i] + 1\n",
    "                sequence_found = True\n",
    "\n",
    "            # in case VL is in covab dab increase the VL count of this entry by 1\n",
    "            if VL_arr[i] in aa_seq:\n",
    "                VL_found[i] = VL_found[i] + 1\n",
    "                sequence_found = True\n",
    "\n",
    "        # sequence that has no match with vh or vl is counted as a not found sequence\n",
    "        if not sequence_found:\n",
    "            sequences_not_in_covabdab = sequences_not_in_covabdab + 1\n",
    "\n",
    "        \n",
    "    return VH_found, VL_found, sequences_not_in_covabdab\n",
    "\n",
    "# prepare all variables for jit\n",
    "sequences_not_in_covabdab = 0\n",
    "VL_arr = CovAbDab['VL'].to_numpy()\n",
    "VH_arr = CovAbDab['VH or VHH'].to_numpy()\n",
    "VH_found = np.zeros((len(VH_arr)))\n",
    "VL_found = np.zeros((len(VL_arr)))\n",
    "aa_seqs = []\n",
    "for i in range(len(proteins)):\n",
    "    aa_seq = Seq(proteins[i]['GBSeq_sequence'])\n",
    "    aa_seqs.append(str.upper(str(aa_seq)))\n",
    "\n",
    "# run function\n",
    "VH_found, VL_found, sequences_not_in_covabdab = search_in_covabdab(aa_seqs, VH_arr, VL_arr, VH_found, VL_found, sequences_not_in_covabdab)\n",
    "\n",
    "# format results\n",
    "sequences_in_covabdab = len(proteins) - sequences_not_in_covabdab\n",
    "CovAbDab_stats_pr = copy.deepcopy(CovAbDab)\n",
    "CovAbDab_stats_pr['VH_found'] = VH_found\n",
    "CovAbDab_stats_pr['VL_found'] = VL_found\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#print summary statistics\n",
    "\n",
    "if True:\n",
    "    print('total sequences assessed:', len(proteins))\n",
    "    print('number of genbank sequences not in covab dab:', sequences_not_in_covabdab)\n",
    "    print('number of genbank sequences found in covab dab:', sequences_in_covabdab)\n",
    "    print('match rate:', sequences_in_covabdab / len(proteins))\n",
    "    # if the total number of counts in VH and VL columns is higher than the genbank sequences that have a match in covab dab\n",
    "    # then a genbank sequence must have several matches in covab dab\n",
    "    print('number of genebank sequences that have multiple matches in covab dab:', (sum(CovAbDab_stats_pr['VH_found'])+sum(CovAbDab_stats_pr['VL_found'])-sequences_in_covabdab))\n",
    "    # if the number of genbank sequences with a match in covab dab is higher than the number of covab dab VH and VLs that were found\n",
    "    # then a number of genbank sequences must have matched to the same covab dab sequenc\n",
    "    # stat is wrong\n",
    "    print('number of genebank entries with non unique match in covab dab:', (sequences_in_covabdab - (len(CovAbDab_stats_pr.loc[(CovAbDab_stats_pr['VH_found'] > 0)]) + len(CovAbDab_stats_pr.loc[(CovAbDab_stats_pr['VL_found'] > 0)]))))\n",
    "    print('-------------')\n",
    "    print('total sequences in covab dab:', len(CovAbDab_stats_pr))\n",
    "    print('number of Covab Dab VH sequences found:', len(CovAbDab_stats_pr.loc[(CovAbDab_stats_pr['VH_found'] > 0)]))\n",
    "    print('number of Covab Dab VL sequences found:', len(CovAbDab_stats_pr.loc[(CovAbDab_stats_pr['VL_found'] > 0)]))\n",
    "    VH_VL_pairings = len(CovAbDab_stats_pr.loc[(CovAbDab_stats_pr['VH_found'] > 0) & (CovAbDab_stats_pr['VL_found'] > 0)])\n",
    "    print('number of Covab Dab VH VL pairings found:', VH_VL_pairings)\n",
    "    VH_or_VL = len(CovAbDab_stats_pr.loc[(CovAbDab_stats_pr['VH_found'] > 0) | (CovAbDab_stats_pr['VL_found'] > 0)])\n",
    "    print('number of Covab Dab entires where either VH or VL was found:', VH_or_VL)\n",
    "    print('-------------')\n",
    "    print('percentage of covab dab entries with pairing found:', VH_VL_pairings / len(CovAbDab_stats_pr) * 100 )\n",
    "    print('percentage of covab dab entries with either VL or VH found:', VH_or_VL / len(CovAbDab_stats_pr) * 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save stats\n",
    "\n",
    "if False:\n",
    "    with open('data/protein_search_stats.csv', 'r+') as fo:\n",
    "        fo.read()\n",
    "        fo.write(f'''{search} - samples: {number_of_samples}, \n",
    "                    {iteration_count}, \n",
    "                    {sequences_not_in_covabdab}, \n",
    "                    {sequences_in_covabdab}, \n",
    "                    {sequences_in_covabdab / iteration_count}, \n",
    "                    {(sum(CovAbDab_stats_pr[\"VH_found\"]) + sum(CovAbDab_stats_pr[\"VL_found\"]) - sequences_in_covabdab)}, \n",
    "                    {(sequences_in_covabdab - (len(CovAbDab_stats_pr.loc[(CovAbDab_stats_pr[\"VH_found\"] > 0)])) + len(CovAbDab_stats_pr.loc[(CovAbDab_stats_pr[\"VL_found\"] > 0)]))}, \n",
    "                    {len(CovAbDab_stats_pr)}, {len(CovAbDab_stats_pr.loc[(CovAbDab_stats_pr[\"VH_found\"] > 0)])}, \n",
    "                    {len(CovAbDab_stats_pr.loc[(CovAbDab_stats_pr[\"VL_found\"] > 0)])}, \n",
    "                    {len(CovAbDab_stats_pr.loc[(CovAbDab_stats_pr[\"VH_found\"] > 0) & (CovAbDab_stats_pr[\"VL_found\"] > 0)])}, \n",
    "                    {len(CovAbDab_stats_pr.loc[(CovAbDab_stats_pr[\"VH_found\"] > 0) | (CovAbDab_stats_pr[\"VL_found\"] > 0)])}, \n",
    "                    {VH_VL_pairings / len(CovAbDab_stats_pr) * 100}, {VH_or_VL / len(CovAbDab_stats_pr) * 100}''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Nucleotide search pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of entries: 2312\n"
     ]
    }
   ],
   "source": [
    "# perform keyword search\n",
    "\n",
    "# specify the terms used for the search\n",
    "search = 'anti-sars-cov-2[All Fields] AND immunoglobulin[All Fields]'\n",
    "\n",
    "# search protein data base with keywords and find out how many entries are found\n",
    "handle = Entrez.esearch(db='nucleotide', term='anti-sars-cov-2[All Fields] AND immunoglobulin[All Fields]', retmax='2')\n",
    "record = Entrez.read(handle)\n",
    "number_of_entries = int(record['Count'])\n",
    "print('number of entries:', number_of_entries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/4195458418.py:16: NumbaWarning: \n",
      "Compilation is falling back to object mode WITH looplifting enabled because Function \"search_in_covabdab\" failed type inference due to: non-precise type array(pyobject, 1d, C)\n",
      "During: typing of argument at /var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/4195458418.py (19)\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/4195458418.py\", line 19:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  @jit\n",
      "/var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/4195458418.py:16: NumbaWarning: \n",
      "Compilation is falling back to object mode WITHOUT looplifting enabled because Function \"search_in_covabdab\" failed type inference due to: Cannot determine Numba type of <class 'numba.core.dispatcher.LiftedLoop'>\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/4195458418.py\", line 22:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  @jit\n",
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/numba/core/object_mode_passes.py:151: NumbaWarning: Function \"search_in_covabdab\" was compiled in object mode without forceobj=True, but has lifted loops.\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/4195458418.py\", line 19:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  warnings.warn(errors.NumbaWarning(warn_msg,\n",
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/numba/core/object_mode_passes.py:161: NumbaDeprecationWarning: \n",
      "Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.\n",
      "\n",
      "For more information visit https://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/4195458418.py\", line 19:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  warnings.warn(errors.NumbaDeprecationWarning(msg,\n",
      "/var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/4195458418.py:16: NumbaWarning: \n",
      "Compilation is falling back to object mode WITHOUT looplifting enabled because Function \"search_in_covabdab\" failed type inference due to: non-precise type array(pyobject, 1d, C)\n",
      "During: typing of argument at /var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/4195458418.py (22)\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/4195458418.py\", line 22:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  @jit\n",
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/numba/core/object_mode_passes.py:151: NumbaWarning: Function \"search_in_covabdab\" was compiled in object mode without forceobj=True.\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/4195458418.py\", line 22:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  warnings.warn(errors.NumbaWarning(warn_msg,\n",
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/numba/core/object_mode_passes.py:161: NumbaDeprecationWarning: \n",
      "Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.\n",
      "\n",
      "For more information visit https://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/4195458418.py\", line 22:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  warnings.warn(errors.NumbaDeprecationWarning(msg,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total sequences assessed: 2312\n",
      "number of genbank sequences not in covab dab: 316\n",
      "number of genbank sequences found in covab dab: 1996\n",
      "match rate: 0.8633217993079585\n",
      "number of genebank sequences that have multiple matches in covab dab: 614.0\n",
      "number of genebank entries with non unique match in covab dab: -55\n",
      "-------------\n",
      "total sequences in covab dab: 4198\n",
      "number of Covab Dab VH sequences found: 1002\n",
      "number of Covab Dab VL sequences found: 1049\n",
      "number of Covab Dab VH VL pairings found: 976\n",
      "number of Covab Dab entires where either VH or VL was found: 1075\n",
      "-------------\n",
      "percentage of covab dab entries with pairing found: 23.249166269652218\n",
      "percentage of covab dab entries with either VL or VH found: 25.607432110528823\n"
     ]
    }
   ],
   "source": [
    "# download all entries from search\n",
    "\n",
    "handle = Entrez.esearch(db='nucleotide', term=search, retmax=number_of_entries)\n",
    "record = Entrez.read(handle)\n",
    "\n",
    "# 25 searches per second\n",
    "nucleotide_handle =  Entrez.efetch(db=\"nucleotide\", id=record['IdList'], rettype=\"gb\", retmode=\"xml\")\n",
    "nucleotides = Entrez.read(nucleotide_handle)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# compare found sequences to covab dab\n",
    "\n",
    "# use numba to speed up computation\n",
    "@jit\n",
    "def search_in_covabdab(aa_seqs, VH_arr, VL_arr, VH_found, VL_found, sequences_not_in_covabdab):\n",
    "\n",
    "    sequences_not_in_covabdab = 0\n",
    "\n",
    "    # loop throught aa seqs\n",
    "    for aa_seq in aa_seqs:\n",
    "        sequence_found = False\n",
    "\n",
    "        # loop throught covab dab entries\n",
    "        for i in range(len(VH_arr)):\n",
    "\n",
    "            # in case VH is in covab dab increase the VH count of this entry by 1\n",
    "            if VH_arr[i] in aa_seq:\n",
    "                VH_found[i] = VH_found[i] + 1\n",
    "                sequence_found = True\n",
    "\n",
    "            # in case VL is in covab dab increase the VL count of this entry by 1\n",
    "            if str(VL_arr[i]) in aa_seq:\n",
    "                VL_found[i] = VL_found[i] + 1\n",
    "                sequence_found = True\n",
    "\n",
    "        # sequence that has no match with vh or vl is counted as a not found sequence\n",
    "        if not sequence_found:\n",
    "            sequences_not_in_covabdab = sequences_not_in_covabdab + 1\n",
    "\n",
    "        \n",
    "    return VH_found, VL_found, sequences_not_in_covabdab\n",
    "\n",
    "# prepare all variables for jit\n",
    "sequences_not_in_covabdab = 0\n",
    "VL_arr = CovAbDab['VL'].to_numpy()\n",
    "VH_arr = CovAbDab['VH or VHH'].to_numpy()\n",
    "VH_found = np.zeros((len(VH_arr)))\n",
    "VL_found = np.zeros((len(VL_arr)))\n",
    "aa_seqs = []\n",
    "for i in range(len(nucleotides)):\n",
    "    nt_seq = Seq(nucleotides[i]['GBSeq_sequence'])\n",
    "    lenght_sequence = len(nt_seq)\n",
    "    remove_bases = lenght_sequence % 3\n",
    "    if remove_bases > 0:\n",
    "        nt_seq = nt_seq[:-remove_bases]\n",
    "    aa_seqs.append(str(nt_seq.translate()))\n",
    "\n",
    "# run function\n",
    "VH_found, VL_found, sequences_not_in_covabdab = search_in_covabdab(aa_seqs, VH_arr, VL_arr, VH_found, VL_found, sequences_not_in_covabdab)\n",
    "\n",
    "# format results\n",
    "sequences_in_covabdab = len(nucleotides) - sequences_not_in_covabdab\n",
    "CovAbDab_stats_nt= copy.deepcopy(CovAbDab)\n",
    "CovAbDab_stats_nt['VH_found'] = VH_found\n",
    "CovAbDab_stats_nt['VL_found'] = VL_found\n",
    "\n",
    "\n",
    "\n",
    "#print statistics\n",
    "\n",
    "if True:\n",
    "    print('total sequences assessed:', len(nucleotides))\n",
    "    print('number of genbank sequences not in covab dab:', sequences_not_in_covabdab)\n",
    "    print('number of genbank sequences found in covab dab:', sequences_in_covabdab)\n",
    "    print('match rate:', sequences_in_covabdab / len(nucleotides))\n",
    "    # if the total number of counts in VH and VL columns is higher than the genbank sequences that have a match in covab dab\n",
    "    # then a genbank sequence must have several matches in covab dab\n",
    "    print('number of genebank sequences that have multiple matches in covab dab:', (sum(CovAbDab_stats_nt['VH_found'])+sum(CovAbDab_stats_nt['VL_found'])-sequences_in_covabdab))\n",
    "    # if the number of genbank sequences with a match in covab dab is higher than the number of covab dab VH and VLs that were found\n",
    "    # then a number of genbank sequences must have matched to the same covab dab sequenc\n",
    "    print('number of genebank entries with non unique match in covab dab:', (sequences_in_covabdab - (len(CovAbDab_stats_nt.loc[(CovAbDab_stats_nt['VH_found'] > 0)]) + len(CovAbDab_stats_nt.loc[(CovAbDab_stats_nt['VL_found'] > 0)]))))\n",
    "    print('-------------')\n",
    "    print('total sequences in covab dab:', len(CovAbDab_stats_nt))\n",
    "    print('number of Covab Dab VH sequences found:', len(CovAbDab_stats_nt.loc[(CovAbDab_stats_nt['VH_found'] > 0)]))\n",
    "    print('number of Covab Dab VL sequences found:', len(CovAbDab_stats_nt.loc[(CovAbDab_stats_nt['VL_found'] > 0)]))\n",
    "    VH_VL_pairings = len(CovAbDab_stats_nt.loc[(CovAbDab_stats_nt['VH_found'] > 0) & (CovAbDab_stats_nt['VL_found'] > 0)])\n",
    "    print('number of Covab Dab VH VL pairings found:', VH_VL_pairings)\n",
    "    VH_or_VL = len(CovAbDab_stats_nt.loc[(CovAbDab_stats_nt['VH_found'] > 0) | (CovAbDab_stats_nt['VL_found'] > 0)])\n",
    "    print('number of Covab Dab entires where either VH or VL was found:', VH_or_VL)\n",
    "    print('-------------')\n",
    "    print('percentage of covab dab entries with pairing found:', VH_VL_pairings / len(CovAbDab_stats_nt) * 100 )\n",
    "    print('percentage of covab dab entries with either VL or VH found:', VH_or_VL / len(CovAbDab_stats_nt) * 100)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save statistics\n",
    "\n",
    "if False:\n",
    "    with open('data/nucleotide_search_stats.csv', 'r+') as fo:\n",
    "        fo.read()\n",
    "        fo.write(f'''{search} - samples: {number_of_samples}, \n",
    "                    {iteration_count}, \n",
    "                    {sequences_not_in_covabdab }, \n",
    "                    {sequences_in_covabdab }, \n",
    "                    {sequences_in_covabdab / iteration_count}, \n",
    "                    {(sum(CovAbDab_stats[\"VH_found\"]) + sum(CovAbDab_stats[\"VL_found\"]) - sequences_in_covabdab)}, \n",
    "                    {(sequences_in_covabdab - (len(CovAbDab_stats.loc[(CovAbDab_stats[\"VH_found\"] > 0)])) + len(CovAbDab_stats.loc[(CovAbDab_stats[\"VL_found\"] > 0)]))}, \n",
    "                    {len(CovAbDab_stats)}, {len(CovAbDab_stats.loc[(CovAbDab_stats[\"VH_found\"] > 0)])}, \n",
    "                    {len(CovAbDab_stats.loc[(CovAbDab_stats[\"VL_found\"] > 0)])}, \n",
    "                    {len(CovAbDab_stats.loc[(CovAbDab_stats[\"VH_found\"] > 0) & (CovAbDab_stats[\"VL_found\"] > 0)])}, \n",
    "                    {len(CovAbDab_stats.loc[(CovAbDab_stats[\"VH_found\"] > 0) | (CovAbDab_stats[\"VL_found\"] > 0)])}, \n",
    "                    {VH_VL_pairings / len(CovAbDab_stats) * 100}, {VH_or_VL / len(CovAbDab_stats) * 100}''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Search both protein and nucelotide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of proteins: 2685\n",
      "number of nucelotides: 2312\n"
     ]
    }
   ],
   "source": [
    "# perform keyword search\n",
    "\n",
    "# specify the terms used for the search\n",
    "search = 'anti-sars-cov-2[All Fields] AND immunoglobulin[All Fields]'\n",
    "\n",
    "# search protein data base with keywords and find out how many entries are found\n",
    "handle_pr = Entrez.esearch(db='protein', term='anti-sars-cov-2[All Fields] AND immunoglobulin[All Fields]', retmax='2')\n",
    "record_pr = Entrez.read(handle_pr)\n",
    "number_of_proteins = int(record_pr['Count'])\n",
    "print('number of proteins:', number_of_proteins)\n",
    "\n",
    "\n",
    "# search protein data base with keywords and find out how many entries are found\n",
    "handle_nt = Entrez.esearch(db='nucleotide', term='anti-sars-cov-2[All Fields] AND immunoglobulin[All Fields]', retmax='2')\n",
    "record_nt = Entrez.read(handle_nt)\n",
    "number_of_nucleotides = int(record_nt['Count'])\n",
    "print('number of nucelotides:', number_of_nucleotides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/3240083081.py:35: NumbaWarning: \n",
      "Compilation is falling back to object mode WITH looplifting enabled because Function \"search_in_covabdab\" failed type inference due to: non-precise type array(pyobject, 1d, C)\n",
      "During: typing of argument at /var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/3240083081.py (39)\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/3240083081.py\", line 39:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  @jit\n",
      "/var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/3240083081.py:35: NumbaWarning: \n",
      "Compilation is falling back to object mode WITHOUT looplifting enabled because Function \"search_in_covabdab\" failed type inference due to: Cannot determine Numba type of <class 'numba.core.dispatcher.LiftedLoop'>\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/3240083081.py\", line 39:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  @jit\n",
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/numba/core/object_mode_passes.py:151: NumbaWarning: Function \"search_in_covabdab\" was compiled in object mode without forceobj=True, but has lifted loops.\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/3240083081.py\", line 39:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  warnings.warn(errors.NumbaWarning(warn_msg,\n",
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/numba/core/object_mode_passes.py:161: NumbaDeprecationWarning: \n",
      "Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.\n",
      "\n",
      "For more information visit https://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/3240083081.py\", line 39:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  warnings.warn(errors.NumbaDeprecationWarning(msg,\n",
      "/var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/3240083081.py:35: NumbaWarning: \n",
      "Compilation is falling back to object mode WITHOUT looplifting enabled because Function \"search_in_covabdab\" failed type inference due to: non-precise type array(pyobject, 1d, C)\n",
      "During: typing of argument at /var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/3240083081.py (39)\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/3240083081.py\", line 39:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  @jit\n",
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/numba/core/object_mode_passes.py:151: NumbaWarning: Function \"search_in_covabdab\" was compiled in object mode without forceobj=True.\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/3240083081.py\", line 39:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  warnings.warn(errors.NumbaWarning(warn_msg,\n",
      "/Users/fabian/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/numba/core/object_mode_passes.py:161: NumbaDeprecationWarning: \n",
      "Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.\n",
      "\n",
      "For more information visit https://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit\n",
      "\n",
      "File \"../../../../../../../../var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/3240083081.py\", line 39:\n",
      "<source missing, REPL/exec in use?>\n",
      "\n",
      "  warnings.warn(errors.NumbaDeprecationWarning(msg,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total sequences assessed: 4997\n",
      "number of genbank sequences not in covab dab: 725\n",
      "number of genbank sequences found in covab dab: 4272\n",
      "match rate: 0.8549129477686612\n",
      "number of genebank sequences that have multiple matches in covab dab: 1250.0\n",
      "number of genebank entries with non unique match in covab dab: 2009\n",
      "-------------\n",
      "total sequences in covab dab: 4198\n",
      "number of Covab Dab VH sequences found based on nucleotide: 1002\n",
      "number of Covab Dab VL sequences found based on nucleotide: 1049\n",
      "number of Covab Dab VH sequences found based on protein: 1110\n",
      "number of Covab Dab VL sequences found based on protein: 1153\n",
      "number of Covab Dab VH sequences found total: 1110\n",
      "number of Covab Dab VL sequences found total: 1153\n",
      "number of Covab Dab VH VL pairings found: 1059\n",
      "number of Covab Dab entires where either VH or VL was found: 1204\n",
      "-------------\n",
      "percentage of covab dab entries with pairing found: 25.226298237255833\n",
      "percentage of covab dab entries with either VL or VH found: 28.680323963792283\n"
     ]
    }
   ],
   "source": [
    "# download all entries from search\n",
    "def download_entries(database):\n",
    "\n",
    "    if database == 'protein':\n",
    "        handle = Entrez.esearch(db='protein', term=search, retmax=number_of_proteins)\n",
    "        record = Entrez.read(handle)\n",
    "\n",
    "        # 25 searches per second\n",
    "        entry_handle =  Entrez.efetch(db=\"protein\", id=record['IdList'], rettype=\"gb\", retmode=\"xml\")\n",
    "        entries = Entrez.read(entry_handle)\n",
    "\n",
    "    if database == 'nucleotide':\n",
    "        handle = Entrez.esearch(db='nucleotide', term=search, retmax=number_of_nucleotides)\n",
    "        record = Entrez.read(handle)\n",
    "\n",
    "        # 25 searches per second\n",
    "        entry_handle =  Entrez.efetch(db=\"nucleotide\", id=record['IdList'], rettype=\"gb\", retmode=\"xml\")\n",
    "        entries = Entrez.read(entry_handle)\n",
    "\n",
    "    search_results[database] = entries\n",
    "\n",
    "\n",
    "search_results = {}\n",
    "for database in ['protein','nucleotide']:\n",
    "    download_entries(database)\n",
    "\n",
    "# run searches on multiprocessing\n",
    "#pool = multiprocessing.Pool()\n",
    "#pool.map(download_entries, ['protein', 'nucleotide'])\n",
    "#pool.close()\n",
    "\n",
    "# compare found sequences to covab dab\n",
    "\n",
    "# use numba to speed up computation\n",
    "@jit\n",
    "def search_in_covabdab(aa_seqs, VH_arr, VL_arr, VH_found, VL_found, sequences_not_in_covabdab):\n",
    "\n",
    "    # loop throught aa seqs\n",
    "    for aa_seq in aa_seqs:\n",
    "        sequence_found = False\n",
    "\n",
    "        # loop throught covab dab entries\n",
    "        for i in range(len(VH_arr)):\n",
    "\n",
    "            # in case VH is in covab dab increase the VH count of this entry by 1\n",
    "            if VH_arr[i] in aa_seq:\n",
    "                VH_found[i] = VH_found[i] + 1\n",
    "                sequence_found = True\n",
    "\n",
    "            # in case VL is in covab dab increase the VL count of this entry by 1\n",
    "            if str(VL_arr[i]) in aa_seq:\n",
    "                VL_found[i] = VL_found[i] + 1\n",
    "                sequence_found = True\n",
    "\n",
    "        # sequence that has no match with vh or vl is counted as a not found sequence\n",
    "        if not sequence_found:\n",
    "            sequences_not_in_covabdab = sequences_not_in_covabdab + 1\n",
    "\n",
    "        \n",
    "    return VH_found, VL_found, sequences_not_in_covabdab\n",
    "\n",
    "# prepare all variables for jit\n",
    "VL_arr = CovAbDab['VL'].to_numpy()\n",
    "VH_arr = CovAbDab['VH or VHH'].to_numpy()\n",
    "VH_found_pr = np.zeros((len(VH_arr)))\n",
    "VL_found_pr = np.zeros((len(VL_arr)))\n",
    "VH_found_nt = np.zeros((len(VH_arr)))\n",
    "VL_found_nt = np.zeros((len(VL_arr)))\n",
    "\n",
    "\n",
    "sequences_not_in_covabdab_nt = 0\n",
    "aa_seqs_nt = []\n",
    "for i in range(len(search_results['nucleotide'])):\n",
    "    nt_seq = Seq(search_results['nucleotide'][i]['GBSeq_sequence'])\n",
    "    lenght_sequence = len(nt_seq)\n",
    "    remove_bases = lenght_sequence % 3\n",
    "    if remove_bases > 0:\n",
    "        nt_seq = nt_seq[:-remove_bases]\n",
    "    aa_seqs_nt.append(str(nt_seq.translate()))\n",
    "\n",
    "# prepare all variables for jit\n",
    "sequences_not_in_covabdab_pr = 0\n",
    "aa_seqs_pr = []\n",
    "for i in range(len(search_results['protein'])):\n",
    "    aa_seq = Seq(search_results['protein'][i]['GBSeq_sequence'])\n",
    "    aa_seqs_pr.append(str.upper(str(aa_seq)))\n",
    "\n",
    "# run function\n",
    "VH_found_nt, VL_found_nt, sequences_not_in_covabdab_nt = search_in_covabdab(aa_seqs_nt, VH_arr, VL_arr, VH_found_nt, VL_found_nt, sequences_not_in_covabdab_nt)\n",
    "VH_found_pr, VL_found_pr, sequences_not_in_covabdab_pr = search_in_covabdab(aa_seqs_pr, VH_arr, VL_arr, VH_found_pr, VL_found_pr, sequences_not_in_covabdab_pr)\n",
    "\n",
    "# format results\n",
    "sequences_in_covabdab_nt = len(search_results['nucleotide']) - sequences_not_in_covabdab_nt\n",
    "sequences_in_covabdab_pr = len(search_results['protein']) - sequences_not_in_covabdab_pr\n",
    "total_sequences = len(search_results['nucleotide']) + len(search_results['protein'])\n",
    "total_sequences_in_covabdab = sequences_in_covabdab_pr + sequences_in_covabdab_nt\n",
    "total_sequences_not_in_covabdab = sequences_not_in_covabdab_nt + sequences_not_in_covabdab_pr\n",
    "CovAbDab_stats= copy.deepcopy(CovAbDab)\n",
    "\n",
    "CovAbDab_stats['VH_found_nucleotide'] = VH_found_nt\n",
    "CovAbDab_stats['VL_found_nucleotide'] = VL_found_nt\n",
    "CovAbDab_stats['VH_found_protein'] = VH_found_pr\n",
    "CovAbDab_stats['VL_found_protein'] = VL_found_pr\n",
    "CovAbDab_stats['VH_found'] = VH_found_pr + VH_found_nt # total\n",
    "CovAbDab_stats['VL_found'] = VL_found_pr + VL_found_nt # total\n",
    "\n",
    "\n",
    "\n",
    "#print statistics\n",
    "\n",
    "if True:\n",
    "    print('total sequences assessed:', total_sequences)\n",
    "    print('number of genbank sequences not in covab dab:', total_sequences_not_in_covabdab)\n",
    "    print('number of genbank sequences found in covab dab:', total_sequences_in_covabdab)\n",
    "    print('match rate:', total_sequences_in_covabdab / total_sequences)\n",
    "    # if the total number of counts in VH and VL columns is higher than the genbank sequences that have a match in covab dab\n",
    "    # then a genbank sequence must have several matches in covab dab\n",
    "    print('number of genebank sequences that have multiple matches in covab dab:', (sum(CovAbDab_stats['VH_found'])+sum(CovAbDab_stats['VL_found'])-total_sequences_in_covabdab))\n",
    "    # if the number of genbank sequences with a match in covab dab is higher than the number of covab dab VH and VLs that were found\n",
    "    # then a number of genbank sequences must have matched to the same covab dab sequenc\n",
    "    print('number of genebank entries with non unique match in covab dab:', (total_sequences_in_covabdab - (len(CovAbDab_stats.loc[(CovAbDab_stats['VH_found'] > 0)]) + len(CovAbDab_stats.loc[(CovAbDab_stats['VL_found'] > 0)]))))\n",
    "    print('-------------')\n",
    "    print('total sequences in covab dab:', len(CovAbDab_stats))\n",
    "    print('number of Covab Dab VH sequences found based on nucleotide:', len(CovAbDab_stats.loc[(CovAbDab_stats['VH_found_nucleotide'] > 0)]))\n",
    "    print('number of Covab Dab VL sequences found based on nucleotide:', len(CovAbDab_stats.loc[(CovAbDab_stats['VL_found_nucleotide'] > 0)]))\n",
    "    print('number of Covab Dab VH sequences found based on protein:', len(CovAbDab_stats.loc[(CovAbDab_stats['VH_found_protein'] > 0)]))\n",
    "    print('number of Covab Dab VL sequences found based on protein:', len(CovAbDab_stats.loc[(CovAbDab_stats['VL_found_protein'] > 0)]))\n",
    "    print('number of Covab Dab VH sequences found total:', len(CovAbDab_stats.loc[(CovAbDab_stats['VH_found'] > 0)]))\n",
    "    print('number of Covab Dab VL sequences found total:', len(CovAbDab_stats.loc[(CovAbDab_stats['VL_found'] > 0)]))\n",
    "    VH_VL_pairings = len(CovAbDab_stats.loc[(CovAbDab_stats['VH_found'] > 0) & (CovAbDab_stats['VL_found'] > 0)])\n",
    "    print('number of Covab Dab VH VL pairings found:', VH_VL_pairings)\n",
    "    VH_or_VL = len(CovAbDab_stats.loc[(CovAbDab_stats['VH_found'] > 0) | (CovAbDab_stats['VL_found'] > 0)])\n",
    "    print('number of Covab Dab entires where either VH or VL was found:', VH_or_VL)\n",
    "    print('-------------')\n",
    "    print('percentage of covab dab entries with pairing found:', VH_VL_pairings / len(CovAbDab_stats) * 100 )\n",
    "    print('percentage of covab dab entries with either VL or VH found:', VH_or_VL / len(CovAbDab_stats) * 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this search the nucleotide search does not find any new sequences compared to the protein search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Version without numpa (takes 3 min / 1000 seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/8x/40q8fgwd2wg9ptnzw_b8cl480000gn/T/ipykernel_92357/1661901097.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     35\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     36\u001b[0m         \u001b[0;31m# in case VL is in covab dab increase the VL count of this entry by 1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 37\u001b[0;31m         \u001b[0;32mif\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mCovAbDab_stats_nt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m9\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32min\u001b[0m \u001b[0maa_seq\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     38\u001b[0m             \u001b[0mCovAbDab_stats_nt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCovAbDab_stats_nt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m             \u001b[0msequence_found\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    920\u001b[0m             \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mis_iterator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    921\u001b[0m             \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_if_callable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 922\u001b[0;31m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_is_scalar_access\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    923\u001b[0m                 \u001b[0;32mwith\u001b[0m \u001b[0msuppress\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mKeyError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mIndexError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    924\u001b[0m                     \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_value\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtakeable\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_takeable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_is_scalar_access\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1475\u001b[0m         \u001b[0;31m# a) avoid getting things via sections and (to minimize dtype changes)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1476\u001b[0m         \u001b[0;31m# b) provide a performant path\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1477\u001b[0;31m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndim\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1478\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1479\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/pandas/_libs/indexing.pyx\u001b[0m in \u001b[0;36mpandas._libs.indexing.NDFrameIndexerBase.ndim.__get__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/auto-db-pipeline/lib/python3.9/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mndim\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    653\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0ma\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_AXIS_ORDERS\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    654\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 655\u001b[0;31m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    656\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mndim\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    657\u001b[0m         \"\"\"\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# copy covabdab dataframe to add stats\n",
    "CovAbDab_stats_nt= copy.deepcopy(CovAbDab)\n",
    "\n",
    "# add columns to count how many times a certain VH or VL was found in the genbank\n",
    "CovAbDab_stats_nt['VH_found'] = 0\n",
    "CovAbDab_stats_nt['VL_found'] = 0\n",
    "\n",
    "sequences_not_in_covabdab = 0\n",
    "iteration_count = 0\n",
    "\n",
    "# loop throught the feched proteins\n",
    "for nucleotide in nucleotides:\n",
    "\n",
    "    iteration_count += 1\n",
    "\n",
    "    nt_seq = nucleotide['GBSeq_sequence']\n",
    "    lenght_sequence = len(nt_seq)\n",
    "    remove_bases = lenght_sequence % 3\n",
    "    \n",
    "    if remove_bases > 0:\n",
    "        nt_seq = nt_seq[:-remove_bases]\n",
    "\n",
    "    nt_seq = Seq(nt_seq)\n",
    "    aa_seq = nt_seq.translate()\n",
    "\n",
    "    sequence_found = False\n",
    "\n",
    "    # loop throught covab dab entries\n",
    "    for i in range(len(CovAbDab_stats_nt)):\n",
    "        \n",
    "        # in case VH is in covab dab increase the VH count of this entry by 1\n",
    "        if CovAbDab_stats_nt.iloc[i,8] in aa_seq:\n",
    "            CovAbDab_stats_nt.iloc[i,-2] = CovAbDab_stats_nt.iloc[i,-2] + 1\n",
    "            sequence_found = True\n",
    "\n",
    "        # in case VL is in covab dab increase the VL count of this entry by 1\n",
    "        if str(CovAbDab_stats_nt.iloc[i,9]) in aa_seq:\n",
    "            CovAbDab_stats_nt.iloc[i,-1] = CovAbDab_stats_nt.iloc[i,-1] + 1\n",
    "            sequence_found = True\n",
    "\n",
    "    # sequence that has no match with vh or vl is counted as a not found sequence\n",
    "    if not sequence_found:\n",
    "        sequences_not_in_covabdab += 1\n",
    "\n",
    "sequences_in_covabdab = iteration_count - sequences_not_in_covabdab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Next tasks\n",
    "1. optimise keywords to find more sequences\n",
    "2. use multiprocessing\n",
    "3. look at proteins found in genbank that are not in covab dab, are these false positives or relevant antibodies missing from covab dab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       MZ751050                 321 bp    mRNA    linear   ROD 17-AUG-2021\n",
      "DEFINITION  Mus musculus clone 15G9/10D2 anti-SARS-CoV-2 spike protein\n",
      "            immunoglobulin light chain variable region mRNA, partial cds.\n",
      "ACCESSION   MZ751050\n",
      "VERSION     MZ751050.1\n",
      "KEYWORDS    .\n",
      "SOURCE      Mus musculus (house mouse)\n",
      "  ORGANISM  Mus musculus\n",
      "            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;\n",
      "            Mammalia; Eutheria; Euarchontoglires; Glires; Rodentia; Myomorpha;\n",
      "            Muroidea; Muridae; Murinae; Mus; Mus.\n",
      "REFERENCE   1  (bases 1 to 321)\n",
      "  AUTHORS   Zhang,G., Wang,A. and Jiang,M.\n",
      "  TITLE     Epitope profiling reveals the critical antigenic determinants in\n",
      "            SARSCoV-2 RBD-based antigen\n",
      "  JOURNAL   Unpublished\n",
      "REFERENCE   2  (bases 1 to 321)\n",
      "  AUTHORS   Zhang,G., Wang,A. and Jiang,M.\n",
      "  TITLE     Direct Submission\n",
      "  JOURNAL   Submitted (09-AUG-2021) College of Animal Science and Veterinary\n",
      "            Medicine, Henan Agricultural University, No. 63, Zhengzhou\n",
      "            Agricultural Road, Zhengzhou, Henan 450002, China\n",
      "COMMENT     ##Assembly-Data-START##\n",
      "            Sequencing Technology :: Sanger dideoxy sequencing\n",
      "            ##Assembly-Data-END##\n",
      "FEATURES             Location/Qualifiers\n",
      "     source          1..321\n",
      "                     /organism=\"Mus musculus\"\n",
      "                     /mol_type=\"mRNA\"\n",
      "                     /db_xref=\"taxon:10090\"\n",
      "                     /clone=\"15G9/10D2\"\n",
      "                     /cell_line=\"SP2/0-derived hybridoma 15G9/10D2\"\n",
      "                     /cell_type=\"hybridoma\"\n",
      "     CDS             <1..>321\n",
      "                     /codon_start=1\n",
      "                     /product=\"anti-SARS-CoV-2 spike protein immunoglobulin\n",
      "                     light chain variable region\"\n",
      "                     /protein_id=\"QYU59157.1\"\n",
      "                     /translation=\"DIVLTQSPASLAVSLGQRATISYRASKSVSTSGYSYMHWNQQKP\n",
      "                     GQPPRLLIYLVSNLESGVPARFSGSGSGTDFTLNIHPVEEEDAATYYCQHIRELTRSE\n",
      "                     GGPSW\"\n",
      "ORIGIN      \n",
      "        1 gacattgtgc tgacacagtc tcctgcttcc ttagctgtat ctctggggca gagggccacc\n",
      "       61 atctcataca gggccagcaa aagtgtcagt acatctggct atagttatat gcactggaac\n",
      "      121 caacagaaac caggacagcc acccagactc ctcatctatc ttgtatccaa cctagaatct\n",
      "      181 ggggtccctg ccaggttcag tggcagtggg tctgggacag acttcaccct caacatccat\n",
      "      241 cctgtggagg aggaggatgc tgcaacctat tactgtcagc acattaggga gcttacacgt\n",
      "      301 tcggaggggg gaccaagctg g\n",
      "//\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#get more infomrtion for each entry\n",
    "handle = Entrez.esearch(db='nucleotide', term='anti-sars-cov-2[All Fields] AND immunoglobulin[All Fields]', retmax='10')\n",
    "record = Entrez.read(handle)\n",
    "\n",
    "genbank_entries_2 = []\n",
    "\n",
    "for ID in record['IdList']:\n",
    "    genbank_entries_2.append(Entrez.efetch(db=\"nucleotide\", id=ID, rettype=\"gb\", retmode=\"text\").read())\n",
    "\n",
    "print(genbank_entries_2[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "f5cb82b301b87eef88325f480de257324ba8b611cc1a51609875e0120daba8a9"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit ('auto-db-pipeline': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
